hw_07.09.23/task2/phonebook.py

```diff
@@ -0,0 +1,138 @@
+import os
+import json
+import argparse
+
+parser = argparse.ArgumentParser()
+parser.add_argument("data_path", help="Phonebook file path")
+args = parser.parse_args()
+
+current_dir = os.path.dirname(os.path.realpath(__file__))
+
+PHONEBOOK = {}
+
+datafile_path = os.path.join(current_dir, (args.data_path + ".json"))
+
+if os.path.exists(datafile_path):
+    with open(datafile_path, "r") as file:
+        PHONEBOOK = json.load(file)
+
+
+def add_contact():
+    first_name = input("Enter first name: ")
+    last_name = input("Enter last name: ")
+    full_name = first_name + " " + last_name
+    telephone_number = input("Enter telephone number: ")
+    city_or_state = input("Enter city or state: ")
+
+    if telephone_number and full_name:
+        PHONEBOOK[telephone_number] = {
+            "first_name": first_name,
+            "last_name": last_name,
+            "full_name": full_name,
+            "address": city_or_state,
+        }
+
+        with open(datafile_path, "w") as file_obj:
+            json.dump(PHONEBOOK, file_obj, indent=4)
+            print(PHONEBOOK)
+
+    else:
+        print("Invalid input. Both name and telephone number are required.")
+
+
+def search_by_fullname(type_: str) -> dict:
+    return {number: contact for number, contact in PHONEBOOK.items() if type_.lower() in contact['full_name'].lower()}
+
+
+def search_by_phone(type_: str) -> dict:
+    return {number: contact for number, contact in PHONEBOOK.items() if type_.lower() in number.lower()}
+
+
+def search_by_city(type_: str) -> dict:
+    return {number: contact for number, contact in PHONEBOOK.items() if type_.lower() in contact['address'].lower()}
+
+
+def update_phonebook(phone):
+    first_name = input("Enter first name: ")
+    last_name = input("Enter last name: ")
+    full_name = first_name + " " + last_name
+    city_or_state = input("Enter city or state: ")
+
+    updated_phonebook = {}
+    if phone and full_name:
+        updated_phonebook[phone] = {
+            "first_name": first_name,
+            "last_name": last_name,
+            "full_name": full_name,
+            "address": city_or_state,
+        }
+        PHONEBOOK.update(updated_phonebook)
+
+        with open(datafile_path, "w") as file_obj:
+            json.dump(PHONEBOOK, file_obj, indent=4)
+            print(PHONEBOOK)
+
+
+def delete_phone():
+    number = input("Enter telephone number to delete: ")
+    if number in PHONEBOOK:
```

Comment:
It's very easy to make a mistake because you rely on the dictionary `PHONEBOOK` that always must be a copy of the phonebook in the file. If at some point you forget to update them synchronously - there will be bugs. It's better to load it every time from the disk (maybe create a function for this purpose). You'll have more operations, but the code will be safer and more robust